### ISSUE TITLE
Tag based IAM policy resource conditions error

### ISSUE THREAD

**Issue by sjparkinson** [hashicorp/terraform#7577]
Creating a IAM policy that uses a resource tag in it's conditions to restrict calls to modification based actions is unsupported in Terraform. This is due to resource tags not being the first action done to newly created resources.

### Terraform Version

Terraform v0.6.16

### Affected Resource(s)

* RDS resources
* EC2 resources

### Terraform Configuration Files

```hcl
variable "vpc" {
    type = "string"
}

resource "aws_security_group" "db_security_group" {
    vpc_id = "${var.vpc}"

    ingress {
        from_port = 3306
        to_port = 3306
        protocol = "tcp"
        cidr_blocks = ["0.0.0.0/0"]
    }

    egress {
        from_port = 0
        to_port = 0
        protocol = "-1"
        cidr_blocks = ["0.0.0.0/0"]
    }

    tags {
        Name = "Terraform Database Security Group"
        System = "Terraform"
    }
}
```

### Debug Output

https://gist.github.com/sjparkinson/b81289cadd412493b58189cf57ee2f9b

### Expected Behavior

Successful creation.

### Actual Behavior

UnauthorizedOperation error when applying the plan.

### Steps to Reproduce

Use the following IAM policy, that uses tag based resource restrictions.

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "ec2:CreateSecurityGroup",
                "ec2:CreateTags",
                "ec2:DeleteTags",
                "ec2:Describe*",
            ],
            "Resource": "*"
        },
        {
            "Effect": "Allow",
            "Action": [
                "ec2:AuthorizeSecurityGroupEgress",
                "ec2:AuthorizeSecurityGroupIngress",
                "ec2:DeleteSecurityGroup",
                "ec2:RevokeSecurityGroupEgress",
                "ec2:RevokeSecurityGroupIngress"
            ],
            "Resource": "*",
            "Condition": {
                "StringEqualsIgnoreCase": {
                    "ec2:ResourceTag/systemCode": "Terraform"
                }
            }
        }
    ]
}
```

### Important Factoids

The tight ACL requires that tags are created for the resource before calling any actions that modify the group.

Currently that doesn't seem to be the case and we have to open up the policy to support Terraform.

**Comment by catsby** [comment]
Hey @sjparkinson interesting ... the problem I'm seeing is that we're setting tags too late:

- https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_security_group.go#L343

That doesn't happen until the end of the update method. We need to move that (and other calls in other resources) up higher, I suppose as close to the creation process as possible. 

I don't know I'll get to this soon, but that's the gist of what needs to be done in the event someone can pick it up

**Comment by sjparkinson** [comment]
> We need to move that (and other calls in other resources) up higher, I suppose as close to the creation process as possible.

That's the jist of it. Currently only RDS and some EC2 actions support this sort of condition, but I suspect it may slowly roll out to others.

**Comment by catsby** [comment]
https://github.com/hashicorp/terraform/pull/7849 should address this by assigning the tags right after creation. We're waiting for another test at the moment but should be merged soon. If you have the chance, can you check it out? If you're not able to build that branch that's fine too, thanks!

**Comment by sjparkinson** [comment]
Sorry for the slow response, I've put in a reminder to check this Sunday.

**Comment by stack72** [comment]
Hi @sjparkinson 

This is going to be merged today so this will be part of the 0.7.1 release - sorry it took so long to follow up with the extra work needed here

P.

**Comment by sjparkinson** [comment]
Thanks a ton!
